src/Abstractions/Entities/TaskEntity.cs

```diff
@@ -71,176 +69,75 @@ public interface ITaskEntity
 ///
 /// <para><b>Entity State</b></para>
 /// <para>
-/// Unchanged from <see cref="ITaskEntity"/>. Entity state is the serialized value of the entity after an operation
-/// completes.
+/// Entity state will be hydrated into the <see cref="TaskEntity{TState}.State"/> property. The contents of this
+/// property will be persisted to <see cref="TaskEntityContext.SetState(object?)"/> when the operation has completed.
+/// Deleting entity state can be accomplished by setting to default(<typeparamref name="TState"/>).
 /// </para>
 /// </remarks>
-public abstract class TaskEntity : ITaskEntity
+public abstract class TaskEntity<TState> : ITaskEntity
 {
-    /**
-     * TODO:
-     * 1. Consider caching a compiled delegate for a given operation name.
-     */
-    static readonly BindingFlags InstanceBindingFlags
-            = BindingFlags.Public | BindingFlags.Instance | BindingFlags.IgnoreCase;
-
-    /// <inheritdoc/>
-    public ValueTask<object?> RunAsync(TaskEntityOperation operation)
-    {
-        Check.NotNull(operation);
-        if (!this.TryDispatchMethod(operation, out object? result, out Type returnType))
-        {
-            throw new NotSupportedException($"No suitable method found for entity operation '{operation}'.");
-        }
-
-        if (typeof(Task).IsAssignableFrom(returnType))
-        {
-            // Task or Task<T>
-            return new(AsGeneric((Task)result!, returnType)); // we assume a declared Task return type is never null.
-        }
-
-        if (returnType == typeof(ValueTask))
-        {
-            // ValueTask
-            return AsGeneric((ValueTask)result!); // we assume a declared ValueTask return type is never null.
-        }
-
-        if (returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(ValueTask<>))
-        {
-            // ValueTask<T>
-            return AsGeneric(result!, returnType); // No inheritance, have to do purely via reflection.
-        }
-
-        return new(result);
-    }
-
-    static bool TryGetInput(ParameterInfo parameter, TaskEntityOperation operation, out object? input)
-    {
-        if (!operation.HasInput)
-        {
-            if (parameter.HasDefaultValue)
-            {
-                input = parameter.DefaultValue;
-                return true;
-            }
-
-            input = null;
-            return false;
-        }
+    /// <summary>
+    /// Gets a value indicating whether dispatching operations to <see cref="State"/> is allowed. State dispatch
+    /// will only be attempted if entity-level dispatch does not succeed. Default is <c>false</c>.
+    /// </summary>
+    protected virtual bool AllowStateDispatch => false;
```

Comment:
@sebastianburckhardt want to draw your attention to this. I set the default to `false` because I feel like that is safer. Don't want users accidentally dispatching to state when they don't realize this property & behavior exists.

Reply:
Setting this to false by default makes sense for the way this interface is currently defined. However, I think there is still a gap here.

I think the dilemma here is about whether we are trying to provide (a) a flexible interface that is constructed logically and can support many different types of `TState` and (b) an interface that aggressively minimizes boilerplate for the most common and simplest use where the user wants to just write the entity just as a simple single class with methods to define the operations, i.e. they only want to look at the class `TState` and have the existence of `TaskEntity<TState>` be an implementation detail that gets hidden in some minimal boilerplate or perhaps even hidden completely at some point.

As currently designed, `TaskEntity<TState>` matches the requirements for (a), but not (b). I realize it may not be possible to satisfy both requirements at the same time..

But we really also need something for (b) because it just seems to matter the most in my experience - it is how we want to write all our samples and benchmarks. I remember that for DF we designed multiple APIs but when I was writing applications it turned out I only ever wanted the single-class one because the code looks so much nicer that way. Having the state indirection is a pretty big hassle IMO, and *forcing* our users to do that is a step backwards (*allowing* them to do it is fine).

So, I still think we need to somehow find a way to make that simple single-class experience with minimal boilerplate possible. Maybe use a derived class  `ClassBasedEntity<TClass>` and then constrain all the choices to keep things simple:
- TClass must be a class with a default constructor
- InitializeState calls the default constructor
- DeleteState remains available on the context so entities can delete themselves
- All operations are dispatched to the inner class first, then the outer class

Reply:
Given the current priorities I would be o.k. with delaying the design and implementation of this "easier to use" single-class version and focus on the more verbose and flexible TaskEntity<T> for now.

Reply:
Per our offline discussion, current experience is sufficient for now. All of these scenarios are possible, but might not be a 1 to 1 with entities in-proc. Which is OK, we can expect some changes during migration. We will revisit this later based on user demand.